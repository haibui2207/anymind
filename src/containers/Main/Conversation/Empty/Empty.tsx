import React, { memo } from "react";

import coverImage from "../../../../assets/images/empty-conversation.jpg";

import useStyles from "./empty.styles";

const Empty: React.FC<{}> = () => {
  const classes = useStyles();

  return (
    <div className={classes.root}>
      <img
        alt="No message"
        className={classes.image}
        src={coverImage}
        data-test="empty-image"
      />
      <h2 className={classes.title} data-test="empty-title">
        No message found
      </h2>
      <p className={classes.description} data-test="empty-description">
        Let's start your conversation
      </p>
    </div>
  );
};

export default memo(Empty);
