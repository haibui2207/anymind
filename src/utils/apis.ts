import { gql } from '@apollo/client';

const FETCH_LATEST_MESSAGES = gql`
  query fetchLatestMessages($channelId: String!) {
    fetchLatestMessages(channelId: $channelId) {
      messageId
      text
      datetime
      userId
    }
  }
`;

const POST_MESSAGE = gql`
  mutation postMessage($channelId: String!, $text: String!, $userId: String!) {
    postMessage(channelId: $channelId, text: $text, userId: $userId) {
      messageId
      text
      datetime
      userId
    }
  }
`;

export default {
  FETCH_LATEST_MESSAGES,
  POST_MESSAGE,
}