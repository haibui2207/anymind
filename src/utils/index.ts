export { default as apiUtil } from './apis'