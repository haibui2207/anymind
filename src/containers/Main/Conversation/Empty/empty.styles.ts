import { createUseStyles } from "react-jss";

import { rem } from "../../../../utils/jss";

export default createUseStyles({
  root: {
    padding: rem(16),
    width: "100%",
    height: "100%",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    flexDirection: "column",
    boxSizing: "border-box",
    textAlign: "center",
  },
  image: { display: "block", width: "100%", maxWidth: rem(250) },
  title: {
    marginTop: rem(16),
    marginBottom: 0,
    fontSize: rem(32),
    lineHeight: rem(48),
  },
  description: {
    marginTop: rem(12),
    marginBottom: 0,
    fontSize: rem(16),
    lineHeight: rem(24),
  },
});
