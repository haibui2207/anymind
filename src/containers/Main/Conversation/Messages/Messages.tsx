import React, { memo, useCallback, useEffect, useMemo } from "react";
import { useLazyQuery } from "@apollo/client";

import { apiUtil } from "../../../../utils";
import { IMessage, IUser } from "../../../../models";
import { useConversationContext } from "../../../../providers/ConversationProvider";

import Message from "./Message";
import useStyles from "./messages.styles";

interface IFormattedMessage extends IMessage {
  user?: IUser;
}

const Messages: React.FC<{}> = () => {
  const classes = useStyles();
  const {
    users,
    messages,
    activeUser,
    activeChannel,
    setMessages,
  } = useConversationContext();
  const [fetchMoreMessages, { called, loading, error, data }] = useLazyQuery<
    IMessage[]
  >(apiUtil.FETCH_MORE_MESSAGES);
  const formattedMessages = useMemo(
    (): IFormattedMessage[] =>
      messages.map((message: IMessage) => ({
        ...message,
        user: users.find((user: IUser) => user.id === message.userId),
      })),
    [messages]
  );

  useEffect((): void => {
    if (called && !loading) {
      if (!error && data) {
        setMessages([...data, ...messages]);
      } else {
        // TODO show error cannot get list messages
        window.alert("Cannot fetch messages");
        console.error(error);
      }
    }
  }, [messages, called, loading, error, data]);

  const handleButtonClick = useCallback((): void => {
    fetchMoreMessages({
      variables: {
        channelId: activeChannel?.id,
        messageId: messages[0].messageId,
        old: true,
      },
    });
  }, [messages, activeChannel]);

  return (
    <div className={classes.root}>
      <button
        className={classes.moreButton}
        onClick={handleButtonClick}
        disabled={loading}
      >
        {loading ? (
          <span>Loading...</span>
        ) : (
          <span>Click to load old messages</span>
        )}
      </button>
      {formattedMessages.map(
        (message: IFormattedMessage): React.ReactElement => (
          <Message
            key={message.messageId}
            {...message}
            isSender={message.userId === activeUser?.id}
          />
        )
      )}
    </div>
  );
};

export default memo(Messages);
