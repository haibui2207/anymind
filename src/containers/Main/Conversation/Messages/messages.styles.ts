import { createUseStyles } from "react-jss";

import { rem } from "../../../../utils/jss";
import COLORS from "../../../../constants/COLORS";

export default createUseStyles({
  root: {},
  moreButton: {
    border: "none",
    textAlign: "center",
    fontSize: rem(14),
    lineHeight: rem(16),
    color: COLORS.textGray,
    fontWeight: "normal",
    background: "transparent",
    display: "block",
    margin: [rem(16), "auto"],
    cursor: "pointer",
    "&:hover": { color: COLORS.textBlack },
  },
});
