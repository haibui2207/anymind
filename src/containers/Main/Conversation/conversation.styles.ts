import { createUseStyles } from "react-jss";

import { rem } from "../../../utils/jss";
import COLORS from "../../../constants/COLORS";

export default createUseStyles({
  root: {
    flex: 1,
    boxSizing: "border-box",
    display: "flex",
    flexDirection: "column",
    "& .ck": {
      "&.ck-editor": { flex: 1 },
      "&.ck-toolbar": { display: "none" },
      "&.ck-editor__main": { height: "100%" },
      "&.ck-editor__editable": {
        height: "100%",
        "& p": { fontSize: rem(14), lineHeight: rem(20) },
      },
      "&.ck-editor__editable_inline": {
        border: "none",
        "&::-webkit-scrollbar": { width: 10, height: 3 },
        "&::-webkit-scrollbar-track, &::-webkit-scrollbar-thumbhover": {
          background: "transparent",
        },
        "&::-webkit-scrollbar-thumb": {
          background: "rgba(178, 178, 178, 0.6)",
          borderRadius: 10,
        },
      },
    },
  },
  header: {
    flex: "none",
    margin: 0,
    padding: rem(16),
    fontSize: rem(24),
    lineHeight: rem(32),
    borderBottom: `1px solid ${COLORS.borderBlue}`,
  },
  content: {
    flex: "1 0 0",
    overflow: "auto",
    borderBottom: `1px solid ${COLORS.borderBlue}`,
  },
  chatBox: {
    flex: "none",
    display: "flex",
    minHeight: rem(40),
    maxHeight: rem(120),
  },
  sendButton: {
    flex: "none",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    height: "100%",
    padding: rem(12),
    width: rem(80),
    cursor: "pointer",
    border: "none",
    color: COLORS.textWhite,
    backgroundColor: COLORS.bgBlack,
    transition: "opacity .2s ease",
    "&:not(:disabled):hover": { opacity: 0.85 },
    "&:disabled": { opacity: 0.5, cursor: "not-allowed" },
  },
});
