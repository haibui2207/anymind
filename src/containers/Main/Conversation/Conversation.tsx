import React, { memo, useCallback, useEffect, useState } from "react";
import { useLazyQuery, useMutation, ApolloError } from "@apollo/client";
import { CKEditor } from "@ckeditor/ckeditor5-react";
import ClassicEditor from "@ckeditor/ckeditor5-build-classic";

import { apiUtil } from "../../../utils";
import { IMessage } from "../../../models";
import { useConversationContext } from "../../../providers/ConversationProvider";

import Empty from "./Empty";
import Messages from "./Messages";
import useStyles from "./conversation.styles";

const Conversation: React.FC<{}> = () => {
  const classes = useStyles();
  const {
    messages,
    draftMessage,
    activeUser,
    activeChannel,
    setMessages,
    setDraftMessage,
  } = useConversationContext();
  const [shownErrorPopup, setErrorPopupStatus] = useState<boolean>(false);

  const handlePostMessageSuccess = useCallback(
    (data: { postMessage: IMessage }): void => {
      const newMessages = [...messages];
      newMessages.push(data.postMessage);
      setDraftMessage(undefined);
      setMessages(newMessages);
      const messagesBox = document.getElementById("messages-box");
      if (messagesBox) {
        messagesBox.scrollTop = messagesBox.scrollHeight;
      }
    },
    [messages]
  );

  const handlePostMessageFailed = useCallback((): void => {
    setErrorPopupStatus(true);
    setTimeout((): void => {
      setErrorPopupStatus(false);
    }, 3000);
  }, []);

  const handleFetchLatestMessagesSuccess = useCallback(
    (data: { fetchLatestMessages: IMessage[] }): void => {
      setMessages([...data.fetchLatestMessages]);
    },
    []
  );

  const handleFetchLatestMessagesFailed = useCallback(
    (error: ApolloError): void => {
      // TODO show error cannot get list messages
      window.alert("Cannot fetch messages");
      console.error(error);
    },
    []
  );

  const [postMessage, { loading: sendingMessage }] = useMutation<{
    postMessage: IMessage;
  }>(apiUtil.POST_MESSAGE, {
    onCompleted: handlePostMessageSuccess,
    onError: handlePostMessageFailed,
  });

  const [fetchLatestMessages, { called, refetch }] = useLazyQuery<{
    fetchLatestMessages: IMessage[];
  }>(apiUtil.FETCH_LATEST_MESSAGES, {
    onCompleted: handleFetchLatestMessagesSuccess,
    onError: handleFetchLatestMessagesFailed,
    notifyOnNetworkStatusChange: true,
  });

  useEffect((): void => {
    if (activeChannel) {
      if (called && refetch) {
        refetch({ channelId: activeChannel.id });
      } else {
        fetchLatestMessages({ variables: { channelId: activeChannel.id } });
      }
    }
  }, [called, activeChannel]);

  // TODO using debounce if necessary
  const handleEditorChange = useCallback(
    (text: string): void => {
      setDraftMessage({
        channelId: activeChannel?.id || "",
        userId: activeUser?.id || "",
        text,
      });
    },
    [activeUser, activeChannel]
  );

  const handleSubmit = useCallback((): void => {
    if (!sendingMessage && activeUser && activeChannel && draftMessage) {
      postMessage({
        variables: {
          channelId: activeChannel.id,
          userId: activeUser.id,
          text: draftMessage.text,
        },
      });
    }
  }, [sendingMessage, activeUser, activeChannel, draftMessage]);

  return (
    <div className={classes.root}>
      <h2 className={classes.header} data-test="conversation-channel-name">
        {activeChannel?.name}
      </h2>
      <div className={classes.content}>
        {messages.length > 0 ? <Messages /> : <Empty />}
        {shownErrorPopup ? (
          <div className={classes.errorPopup}>Cannot send the message</div>
        ) : null}
      </div>
      <div className={classes.chatBox}>
        <CKEditor
          editor={ClassicEditor}
          data={draftMessage?.text || ""}
          onChange={(event, editor) => {
            const data = editor.getData();
            handleEditorChange(data);
          }}
          config={{
            placeholder: "Type your message here...",
          }}
        />
        <button
          className={classes.sendButton}
          title="Send message"
          disabled={sendingMessage || !draftMessage?.text}
          onClick={handleSubmit}
          data-test="conversation-submit-button"
        >
          <i className="icon-plane" />
        </button>
      </div>
    </div>
  );
};

export default memo(Conversation);
