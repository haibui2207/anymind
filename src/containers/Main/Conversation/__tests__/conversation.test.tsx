import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { act } from "react-dom/test-utils";
import { MockedProvider } from "@apollo/client/testing";
import { CKEditor } from "@ckeditor/ckeditor5-react";

import {
  IConversation,
  ConversationContext,
  defaultContextValue,
} from "../../../../providers/ConversationProvider";
import { apiUtil } from "../../../../utils";
import Empty from "../Empty";
import Messages from "../Messages";
import Conversation from "../Conversation";

const waitForComponentUpdate = async (wrapper: any) => {
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 20));
    wrapper.update();
  });
};

describe("Testing Main > Conversation", (): void => {
  let wrapper: any;
  const mockContextValue: IConversation = {
    ...defaultContextValue,
    users: [{ id: "1", name: "User 1", avatar: "" }],
    channels: [{ id: "1", name: "Channel 1" }],
    activeUser: { id: "1", name: "User 1", avatar: "" },
    activeChannel: { id: "1", name: "Channel 1" },
    messages: [],
    setMessages: jest.fn(),
  };
  const mocks = [
    {
      request: {
        query: apiUtil.FETCH_LATEST_MESSAGES,
        variables: { channelId: "1" },
      },
      result: {
        data: {
          fetchLatestMessages: [
            {
              messageId: "1687680749017235835",
              text: "<p>done</p>",
              datetime: "2021-09-30T02:30:00Z",
              userId: "Joyse",
              __typename: "Message",
            },
          ],
        },
      },
    },
    {
      request: {
        query: apiUtil.FETCH_MORE_MESSAGES,
        variables: {
          channelId: "1",
          messageId: "5728579392774769867",
          old: false,
        },
      },
      result: {
        data: {
          fetchMoreMessages: [
            {
              messageId: "1687680749017235835",
              text: "<p>done</p>",
              datetime: "2021-09-30T02:30:00Z",
              userId: "Joyse",
              __typename: "Message",
            },
          ],
        },
      },
    },
    {
      request: {
        query: apiUtil.POST_MESSAGE,
        variables: {
          channelId: "1",
          userId: "Russell",
          text: "Text from Russell",
        },
      },
      result: {
        data: {
          postMessage: [
            {
              messageId: "1687680749017235835",
              text: "Text from Russell",
              datetime: "2021-09-30T02:30:00Z",
              userId: "Russell",
              __typename: "Message",
            },
          ],
        },
      },
    },
  ];

  beforeAll(
    async (): Promise<void> => {
      wrapper = mount(
        <ConversationContext.Provider value={mockContextValue}>
          <MockedProvider mocks={mocks}>
            <Conversation />
          </MockedProvider>
        </ConversationContext.Provider>
      );
      await waitForComponentUpdate(wrapper);
    }
  );

  it("should render match snapshot", (): void => {
    const wrapper = shallow(
      <ConversationContext.Provider value={mockContextValue}>
        <MockedProvider mocks={mocks}>
          <Conversation />
        </MockedProvider>
      </ConversationContext.Provider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should show current active channel name", (): void => {
    expect(
      wrapper.find('[data-test="conversation-channel-name"]')
    ).toHaveLength(1);
  });

  it("should show Empty component when messages empty", async (): Promise<void> => {
    const wrapper = mount(
      <ConversationContext.Provider
        value={{ ...mockContextValue, messages: [] }}
      >
        <MockedProvider mocks={mocks}>
          <Conversation />
        </MockedProvider>
      </ConversationContext.Provider>
    );
    await waitForComponentUpdate(wrapper);
    expect(wrapper.find(Empty)).toHaveLength(1);
    expect(wrapper.find(Messages)).toHaveLength(0);
  });

  it("should show Conversation component when messages exists", async (): Promise<void> => {
    const wrapper = mount(
      <ConversationContext.Provider
        value={{
          ...mockContextValue,
          messages: [
            {
              messageId: "1",
              text: "The first message",
              datetime: "2021-09-29T01:39:08.665Z",
              userId: "Sam",
            },
          ],
        }}
      >
        <MockedProvider mocks={mocks}>
          <Conversation />
        </MockedProvider>
      </ConversationContext.Provider>
    );
    await waitForComponentUpdate(wrapper);
    expect(wrapper.find(Empty)).toHaveLength(0);
    expect(wrapper.find(Messages)).toHaveLength(1);
  });

  it("should have a n Editor", (): void => {
    expect(wrapper.find(CKEditor)).toHaveLength(1);
  });

  it("should have a submit button", (): void => {
    expect(
      wrapper.find('[data-test="conversation-submit-button"]')
    ).toHaveLength(1);
  });
});
