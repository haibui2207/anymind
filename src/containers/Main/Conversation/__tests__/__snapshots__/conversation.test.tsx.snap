// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Main > Conversation should render match snapshot 1`] = `
<MockedProvider
  addTypename={true}
  mocks={
    Array [
      Object {
        "request": Object {
          "query": Object {
            "definitions": Array [
              Object {
                "directives": Array [],
                "kind": "OperationDefinition",
                "name": Object {
                  "kind": "Name",
                  "value": "fetchLatestMessages",
                },
                "operation": "query",
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "channelId",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "channelId",
                            },
                          },
                        },
                      ],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "fetchLatestMessages",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "messageId",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "text",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "datetime",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "userId",
                            },
                            "selectionSet": undefined,
                          },
                        ],
                      },
                    },
                  ],
                },
                "variableDefinitions": Array [
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "channelId",
                      },
                    },
                  },
                ],
              },
            ],
            "kind": "Document",
            "loc": Object {
              "end": 166,
              "start": 0,
            },
          },
          "variables": Object {
            "channelId": "1",
          },
        },
        "result": Object {
          "data": Object {
            "fetchLatestMessages": Array [
              Object {
                "__typename": "Message",
                "datetime": "2021-09-30T02:30:00Z",
                "messageId": "1687680749017235835",
                "text": "<p>done</p>",
                "userId": "Joyse",
              },
            ],
          },
        },
      },
      Object {
        "request": Object {
          "query": Object {
            "definitions": Array [
              Object {
                "directives": Array [],
                "kind": "OperationDefinition",
                "name": Object {
                  "kind": "Name",
                  "value": "fetchMoreMessages",
                },
                "operation": "query",
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "channelId",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "channelId",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "messageId",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "messageId",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "old",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "old",
                            },
                          },
                        },
                      ],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "fetchMoreMessages",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "messageId",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "text",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "datetime",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "userId",
                            },
                            "selectionSet": undefined,
                          },
                        ],
                      },
                    },
                  ],
                },
                "variableDefinitions": Array [
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "channelId",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "messageId",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "Boolean",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "old",
                      },
                    },
                  },
                ],
              },
            ],
            "kind": "Document",
            "loc": Object {
              "end": 233,
              "start": 0,
            },
          },
          "variables": Object {
            "channelId": "1",
            "messageId": "5728579392774769867",
            "old": false,
          },
        },
        "result": Object {
          "data": Object {
            "fetchMoreMessages": Array [
              Object {
                "__typename": "Message",
                "datetime": "2021-09-30T02:30:00Z",
                "messageId": "1687680749017235835",
                "text": "<p>done</p>",
                "userId": "Joyse",
              },
            ],
          },
        },
      },
      Object {
        "request": Object {
          "query": Object {
            "definitions": Array [
              Object {
                "directives": Array [],
                "kind": "OperationDefinition",
                "name": Object {
                  "kind": "Name",
                  "value": "postMessage",
                },
                "operation": "mutation",
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "channelId",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "channelId",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "text",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "text",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "userId",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "userId",
                            },
                          },
                        },
                      ],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "postMessage",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "messageId",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "text",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "datetime",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "userId",
                            },
                            "selectionSet": undefined,
                          },
                        ],
                      },
                    },
                  ],
                },
                "variableDefinitions": Array [
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "channelId",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "text",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "userId",
                      },
                    },
                  },
                ],
              },
            ],
            "kind": "Document",
            "loc": Object {
              "end": 217,
              "start": 0,
            },
          },
          "variables": Object {
            "channelId": "1",
            "text": "Text from Russell",
            "userId": "Russell",
          },
        },
        "result": Object {
          "data": Object {
            "postMessage": Array [
              Object {
                "__typename": "Message",
                "datetime": "2021-09-30T02:30:00Z",
                "messageId": "1687680749017235835",
                "text": "Text from Russell",
                "userId": "Russell",
              },
            ],
          },
        },
      },
    ]
  }
>
  <Memo(Conversation) />
</MockedProvider>
`;
